packages/fetchai/connections/p2p_libp2p/dht/dhtclient/dhtclient.go

```diff
@@ -343,34 +376,99 @@ func (dhtClient *DHTClient) RouteEnvelope(envel *aea.Envelope) error {
 	ldebug().
 		Str("op", "route").
 		Str("target", target).
-		Msg("requesting peer ID from relay...")
+		Msg("requesting agent record from relay...")
 
-	err = utils.WriteBytes(stream, []byte(target))
+	lookupRequest := &dhtnode.LookupRequest{AgentAddress: target}
+	msg := &dhtnode.AcnMessage{
+		Version: dhtnode.CurrentVersion,
+		Payload: &dhtnode.AcnMessage_LookupRequest{LookupRequest: lookupRequest},
+	}
+	buf, err := proto.Marshal(msg)
+	if err != nil {
+		lerror(err).
+			Str("op", "route").
+			Str("target", target).
+			Msg("while serializing LookupRequest message")
+		errReset := stream.Reset()
+		ignore(errReset)
+		return err
+	}
+	err = utils.WriteBytes(stream, buf)
 	if err != nil {
 		lerror(err).
 			Str("op", "route").
 			Str("target", target).
-			Msg("while sending address to relay")
-		return errors.New("ERROR route - While sending address to relay:" + err.Error())
+			Msg("while sending LookupRequest message")
+		errReset := stream.Reset()
+		ignore(errReset)
+		return err
 	}
 
-	msg, err := utils.ReadString(stream)
+	buf, err = utils.ReadBytes(stream)
 	if err != nil {
 		lerror(err).
 			Str("op", "route").
 			Str("target", target).
-			Msgf("while reading target peer id from relay")
-		return errors.New("ERROR route - While reading target peer id from relay:" + err.Error())
+			Msg("while receiving response acn message")
+		errReset := stream.Reset()
+		ignore(errReset)
+		return err
+	}
+
+	response := &dhtnode.AcnMessage{}
+	err = proto.Unmarshal(buf, response)
+	if err != nil {
+		lerror(err).Str("op", "route").Str("target", target).
+			Msgf("couldn't deserialize agent lookup response")
+		return err
+	}
+
+	// response is either a LookupResponse or Status
+	var lookupResponse *dhtnode.LookupResponse = nil
+	var status *dhtnode.Status = nil
+	switch pl := response.Payload.(type) {
+	case *dhtnode.AcnMessage_LookupResponse:
+		lookupResponse = pl.LookupResponse
+	case *dhtnode.AcnMessage_Status:
+		status = pl.Status
+	default:
+		err = errors.New("Unexpected Acn Message")
+		lerror(err).Str("op", "route").Str("target", target).
+			Msgf("couldn't deserialize agent lookup response")
+		return err
+	}
+
+	if status != nil {
+		err = errors.New(status.Code.String() + " : " + strings.Join(status.Msgs, ":"))
+		lerror(err).Str("op", "route").Str("target", target).
+			Msgf("failed agent lookup")
+		return err
+	}
+
+	// lookupResponse must be set
+	record := lookupResponse.AgentRecord
+	valid, err := dhtnode.IsValidProofOfRepresentation(record, target, record.PeerPublicKey)
```

Comment:
[STP3.a] when a peer receives outgoing envelope it proceed to continuously lookup the dht for the destination until it finds a peer with a valid PoR, then forward the envelope and wait for delivery confirmation